src/free_list.h

```diff
@@ -0,0 +1,210 @@
+    return;
+  Log(kCrash, __FILE__, __LINE__, "Circular loop in list detected: ", next);
+}
+
+inline void* MaskPtr(void* p) {
+  // Maximize ASLR entropy and guarantee the result is an invalid address.
+  const uintptr_t mask =
+      ~(reinterpret_cast<uintptr_t>(TCMalloc_SystemAlloc) >> 13);
+  return reinterpret_cast<void*>(reinterpret_cast<uintptr_t>(p) ^ mask);
+}
+
+inline void* UnmaskPtr(void* p) {
+  return MaskPtr(p);
+}
+
+// Returns value of the |previous| pointer w/out running a sanity
+// check.
+inline void* FL_Previous_No_Check(void* t) {
+  return UnmaskPtr(reinterpret_cast<void**>(t)[1]);
+}
+
+// Returns value of the |next| pointer w/out running a sanity check.
+inline void* FL_Next_No_Check(void* t) {
+  return UnmaskPtr(reinterpret_cast<void**>(t)[0]);
+}
+
+inline void* FL_Previous(void* t) {
+  void* previous = FL_Previous_No_Check(t);
+  if (previous) {
+    FL_EqualityCheck(FL_Next_No_Check(previous), t, __FILE__, __LINE__);
+  }
+  return previous;
+}
+
+inline void FL_SetPrevious(void* t, void* n) {
+  EnsureNonLoop(t, n);
+  reinterpret_cast<void**>(t)[1] = MaskPtr(n);
+}
+
+inline void FL_SetNext(void* t, void* n) {
+  EnsureNonLoop(t, n);
+  reinterpret_cast<void**>(t)[0] = MaskPtr(n);
+}
+
+inline void* FL_Next(void* t) {
+  void* next = FL_Next_No_Check(t);
+  if (next) {
+    FL_EqualityCheck(FL_Previous_No_Check(next), t, __FILE__, __LINE__);
+  }
+  return next;
+}
+
+// Pops the top element off the linked list whose first element is at
+// |*list|, and updates |*list| to point to the next element in the
+// list.  Returns the address of the element that was removed from the
+// linked list.  |list| must not be NULL.
+inline void* FL_Pop(void** list) {
+  void* result = *list;
+  ASSERT(FL_Previous_No_Check(result) == NULL);
+  *list = FL_Next(result);
+  if (*list != NULL) {
+    FL_SetPrevious(*list, NULL);
+  }
+  return result;
+}
+
+// Makes the element at |t| a singleton doubly linked list.
+inline void FL_Init(void* t) {
+  FL_SetPrevious(t, NULL);
+  FL_SetNext(t, NULL);
+}
+
+// Pushes element to a linked list whose first element is at
+// |*list|. When this call returns, |list| will point to the new head
+// of the linked list.
+inline void FL_Push(void** list, void* element) {
+  void* old = *list;
+  if (old == NULL) {  // Builds singleton list.
+    FL_Init(element);
+  } else {
+    ASSERT(FL_Previous_No_Check(old) == NULL);
+    FL_SetNext(element, old);
+    FL_SetPrevious(old, element);
+    FL_SetPrevious(element, NULL);
+  }
+  *list = element;
+}
+
+#else  // TCMALLOC_USE_DOUBLYLINKED_FREELIST not defined
+static const bool kSupportsDoublyLinkedList = false;
+
+inline void* FL_Next(void* t) {
+  return SLL_Next(t);
+}
+
+inline void FL_Init(void* t) {
+  SLL_SetNext(t, NULL);
+}
+
+inline void FL_Push(void** list, void* element) {
+  if (*list != element) {
+    SLL_Push(list, element);
+    return;
+  }
+  Log(kCrash, __FILE__, __LINE__, "Double Free of %p detected", element);
```

Comment:
This hopelessly cripples fast-path of free. Extra branch might be ok (I am not sure about that), but this Log call actually cripples code gen too much. I think it could be fixed by ensuring compiler is aware it is noreturn and making sure we have ABSL_PREDICT_XXX properly setup.

I.e. ./malloc_bench first benchmark is 5.1 ns before and 6.5 ns after on my box.

Reply:
Also %p isn't meaningful for this api.